Applicant can edit their Membership Application

tochman opened the pull request:
PT Story: https://www.pivotaltracker.com/story/show/134078325

_As an applicant
In order to be able to edit my application
I want to be allowed to do that_


Changes proposed in this pull request:
1. adds User model with Devise for authentication
2. adds relation User -> MembershipApplication
3. adds Edit view for MembershipApplication
4. extracts form fields to partial
5. introduces authorization engine (Pundit)


Ready for review @thesuss @weedySeaDragon @diraulo et al

tochman committed:
adds scenario for editing submitted application
adds basic step definitions
adds setup for user authentication in tests env

tochman committed:
adds User model and authentication library (Devise)
adds association User -> MembershipApplication

tochman committed:
adds edit view and method
atm we only allow to edit the LAST application a user has submitted

tochman committed:
refactors scenarios and controller methods affected by new functionality

tochman committed:
merges with upstream

tochman committed:
rafactoring after merge conflicts
consolidates step defs to context specific files
refactores create application step and background scenario

weedySeaDragon commented on features/show_membership_application.feature:
```diff
@@ -3,21 +3,27 @@ Feature: As an Admin
-  | Hunderiet    | 1234567890     | Emma Svensson  | 1234-234567  | min@mail.se   |
-  | DoggieZone   | 2345678901     | Pam Andersson  | 0234-234567  | din@mail.se   |
-  | Tassa-in AB  | 1234367890     | Anna Knutsson  | 1234-234569  | sin@mail.se   |
+  Background:
```
This hasn't been implemented yet.  So don't we first need a feature that does this?

tochman replied:
It was implemented in https://github.com/AgileVentures/shf-project/pull/13

weedySeaDragon replied:
`"Given the following user exists..."`  hasn't been implemented yet.  #13 was about the membership application, not users & devise.

But given your comment in the overall conversation about this,  I understand now.  (or at least I'm less confused :-) )

weedySeaDragon commented:
Where does the visitor log in so they become the `current_user`?  

This is the part I haven't been clear about:  This PR seems to assume that a visitor has received their login credentials (user id & password), and that there are the UI pieces in place so that a visitor can log in.  Specifically, the Background `Given the following users exists ...`   has not yet been implemented.  

Don't we need a feature (or features) for this? _(A visitor receives their login credentials.  A visitor can log in)_

thesuss commented:
It needs to be tested too, yes. And prettified. But both could be done in a separate feature. I agree that it's missing, but it's also not critical at the moment. 

That's how I have understood it.

weedySeaDragon commented:
When I run the feature tests, they pass because the `Background` steps have been automagically made to happen.  But if I try to do those same steps manually, I can't.  _There's no way to do them manually._

I understand _why_ you would want to separate the testing such that you have the background steps automagically take care of separate functionality. (= Keeping this functionality/PR well bounded)   But at some point, somewhere in the system the user has to actually be able to do those things: _get log in credentials_ and _log in_

I'm either missing something about how the users would use the system, or I'm not following something about how this PR/user story relates to other user stories in Pivotal Tracker -- or both.  

(And if it's better to move this discussion somewhere else, just let me know.)

tochman commented:
That is correct. There is no way to test this in the browser/manually at the moment. Well, there is, but it is not easy. There are no links to Login in or Log out or Register pages. But the routes are there. Run `rails routes` to see what url's are part of the system. You can then enter those paths manually in the browser window.

tochman committed:
adds Pundit for authorization
keeps policies in ApplicationPolicy. MembershipApplicationPolicy inherits from that
refactors step definitions
adds a logged in user to create application scenario to keep functionality
this is a temp solution until an system admin role is implemented.
refactores rails_helper to make usage of FactoryGirl more DRY

tochman committed:
removes unused argument in application_ontroller#user_not_authorized method

weedySeaDragon commented:
I think I'm getting it now.  :-)  

My confusions stems from the fact that:
1.  There are no explicit features in Pivotal Tracker for doing the UI for a user logging in, nor is there anything about how the user will get their credentials. I think we all understand that must happen, but it hasn't been made explicit.    I'm not yet clear on why those aren't explicit features in Pivotal Tracker, but you all are clearly good with how things are.  So I'll wait and watch and learn :-) .
2. I'm used to bootstrapping an app by having these typical 'login, get credentials...' done quite early.  The client reviews the basic _getting credentials & logging in_ while the first domain functionality is being done. (It's more of a project management decision to do it that way.)  That's _not_ a 'better' way; it's just the way that was stuck in my head, so I had a hard time seeing _this_ way.

tochman committed:
Merge branch 'develop' of github.com:AgileVentures/shf-project into edit_application

tochman committed:
removes unused application_job and application_mailer
removes unused scaffolded policy methods

tochman commented on app/views/partials/_top.html.haml:
```diff
@@ -21,7 +21,10 @@
               = link_to 'Mer', 'http://sverigeshundforetagare.se/'
             %li.page_item.page-item-2
-              = link_to 'Apply for membership', new_membership_path
+              - if current_user.try(:membership_applications).try(:any?) && current_user.try(:membership_applications).try(:last).try(:persisted?)
```
This line is up for debate. This is a temp solution. We would probably want to use Pundit or some helper to check for this. I would argue that this is a refactoring ticket candidate. But I don't know.

thesuss replied:
Lets have it like that for now. We can ignore the hound barks until we have figured out a better setup, and then make a PR with that change.

tochman replied:
I'm okay with that approach

thesuss commented on features/create_membership_application.feature:
```diff
@@ -4,6 +4,12 @@ Feature: As a visitor
 
   PT Feature: https://www.pivotaltracker.com/story/show/133940725
```
Could you please make this on line four and just PT: link ? 
The second instance of the word "Feature" gets coloured in some IDE's and that could be confusing.

thesuss commented on features/edit_membership_application.feature:
```diff
@@ -0,0 +1,31 @@
+  In order to be able to edit my application
+  I want to be allowed to do that
+
+  PT Feature: https://www.pivotaltracker.com/story/show/134078325
```
Could you please make this on line four and just PT: link ? 
The second instance of the word "Feature" gets coloured in some IDE's and that could be confusing.

tochman replied:
Okay.

tochman committed:
updates feature description to avoid reuse of word 'Feature'

tochman commented:
Guys, this feature wasn't complete. I was a bit stressed and forgot to update the `edit` and `update` actions as well as the scenario in `edit_membership_application.feature:19` There is no assertion in that scenario and we have no way of telling if the Company Name actually gets updated to "A Doggy Dog World" (it doesn't).

Lessons learned: 
1. Don't code with small children present
2. Don't trust that your teammates actually read through the acceptance tests when doing code review ;--)

Action points for @thesuss : Make this a part of the ongoing work with setting the status of an application.
